### DISCUSSION TITLE
Why does tensor indexing x[i][j][k] return a different value as x[i,j,k] ?

### DISCUSSION THREAD

**Discussion by shantanu-ghodgaonkar** [mrdbourke/pytorch-deep-learning#816] [Q&A] [answered]
Here is my code - 
```
x = torch.arange(1,10).reshape(1,3,3)
print(f"x = {x}")
print(f"x.shape = {x.shape}")
print(f"x[0][:][2] = {x[0][:][2]}") # returns [7,8,9]
print(f"x[0,:,2] = {x[0,:,2]}") # returns [3,6,9]
```

This is what my output looks like - 
```
x = tensor([[[1, 2, 3],
         [4, 5, 6],
         [7, 8, 9]]])
x.shape = torch.Size([1, 3, 3])
x[0] = tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])
x[0][0] = tensor([1, 2, 3])
x[0][:][2] = tensor([7, 8, 9])
x[0,:,2] = tensor([3, 6, 9])
```

Shouldn't both `x[0][:][2]` and `x[0,:,2]` give the same output?

**Comment by badal-kumar-sah** [comment]
Using [:] is like using a wild character or in simple language using None in python.

Thus, x[:] means selecting all and not decreasing dimensionally.

<img width="443" alt="Screenshot 2024-01-30 at 4 47 06 PM" src="https://github.com/mrdbourke/pytorch-deep-learning/assets/63067043/7babeb03-733c-4420-b1c5-6638ec22eea7">

**Comment by holt5301** [comment, marked as answer]
The two notations do not have the same meaning, though can look like they are the same with some usages.

With `x[i][j][k]` you should think about it as indexing into a nested list of arrays, while the `x[i, j, k]` is more like specifying what ranges you would like in each dimension for a tensor.  At first thought these seem equivalent but they are not.

Lets take away the 3rd dimension from your tensor for simplicity, so now you have a more traditional matrix.

In this notation with indexing as a sequence of operations (e.g. x[:][2]) you can see that x[:] is exactly the same as x, so x[:][2] is the exact same as x[2].  This means you're getting the 3rd array (index 2) in the array of arrays.

```python
In [35]: mat
Out[35]: 
array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [36]: mat[:]
Out[36]: 
array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [37]: mat[:][2]
Out[37]: array([7, 8, 9])

In [38]: mat[2]
Out[38]: array([7, 8, 9])
```

In this example we show that the comma based notation is more like a linear algebra representation that indicates which portion of the matrix is to be returned, first specifying rows and columns.  Using `x[:,2]` means we want all rows, and only the second column.

```python
In [39]: mat[:,:]
Out[39]: 
array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [40]: mat[:,:2]
Out[40]: 
array([[1, 2],
       [4, 5],
       [7, 8]])

In [41]: mat[:,2]
Out[41]: array([3, 6, 9])
```
`

**Reply by shantanu-ghodgaonkar** [reply]
This makes sense, thanks! :)
